gemini/use-cases/retrieval-augmented-generation/qna_using_query_routing/utils/qna_vector_search.py

```diff
@@ -0,0 +1,122 @@
+# Copyright 2023 Google LLC. This software is provided as-is, without warranty
+# or representation for any use or purpose. Your use of it is subject to your
+# agreement with Google.
+"""Answer QnA Type Questions using genai Content"""
+
+# Utils
+import configparser
+import logging
+
+from google.cloud import aiplatform
+import pandas as pd
+from utils import qna_using_query_routing_utils
+from vertexai.generative_models import GenerativeModel
+from vertexai.language_models import TextEmbeddingModel
+
+
+class QnAVectorSearch:
+    """genai Generate Answer From genai Content"""
+
+    def __init__(
+        self, config_file: str = "config.ini", logger=logging.getLogger()
+    ) -> None:
+        self.config = configparser.ConfigParser()
+        self.config.read(config_file)
+        self.project_id = self.config["default"]["project_id"]
+        self.region = self.config["default"]["region"]
+        self.me_region = self.config["vector_search"]["me_region"]
+        self.me_index_name = self.config["vector_search"]["me_index_name"]
+        self.me_gcs_bucket = self.config["vector_search"]["me_gcs_bucket"]
+        self.number_of_references_to_summarise = int(
+            self.config["genai_qna"]["number_of_references_to_summarise"]
+        )
+        self.search_distance_threshould = float(
+            self.config["genai_qna"]["search_distance_threshould"]
+        )
+        self.logger = logger
+
+        # Default retrieval prompt template
+        self.prompt_template = """
+        SYSTEM: You are genai Programming Language Learning Assistant helping the students answer their questions based on following context. Explain the answers in detail for students.
+
+        Instructions:
+        1. Think step-by-step and then answer.
+        2. Explain the answer in detail.
+        3. If the answer to the question cannot be determined from the context alone, say "I cannot determine the answer to that."
+        4. If the context is empty, just say "I could not find any references that are directly related to your question."
+
+        Context:
+        =============
+        {context}
+        =============
+
+        What is the Detailed explanation of answer of following question?
+        Question: {question}
+        Detailed explanation of Answer:"""  # pylint: disable=line-too-long
```

Comment:
There's a few spelling/grammar/capitalization errors. Can you fix them?

Reply:
corrected the prompt and spellings.